engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeAnyNode.java

```diff
@@ -0,0 +1,349 @@
+      },
+      limit = "3")
+  long hashCodeForDateInterop(Object selfDate, @CachedLibrary("selfDate") InteropLibrary interop) {
+    try {
+      return interop.asDate(selfDate).hashCode();
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(
+      guards = {
+        "interop.isDuration(selfDuration)",
+      },
+      limit = "3")
+  long hashCodeForDurationInterop(
+      Object selfDuration, @CachedLibrary("selfDuration") InteropLibrary interop) {
+    try {
+      return interop.asDuration(selfDuration).hashCode();
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization
+  long hashCodeForText(Text text, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    if (text.is_normalized()) {
+      return text.toString().hashCode();
+    } else {
+      return hashCodeForString(text, interop);
+    }
+  }
+
+  @TruffleBoundary
+  @Specialization(
+      guards = {"interop.isString(selfStr)"},
+      limit = "3")
+  long hashCodeForString(Object selfStr, @CachedLibrary("selfStr") InteropLibrary interop) {
+    String str;
+    try {
+      str = interop.asString(selfStr);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+    Normalizer2 normalizer = Normalizer2.getNFDInstance();
+    if (normalizer.isNormalized(str)) {
+      return str.hashCode();
+    } else {
+      return normalizer.normalize(str).hashCode();
+    }
+  }
+
+  @Specialization(
+      guards = {"interop.hasArrayElements(selfArray)"},
+      limit = "3")
+  long hashCodeForArray(
+      Object selfArray,
+      @CachedLibrary("selfArray") InteropLibrary interop,
+      @Cached HashCodeAnyNode hashCodeNode,
+      @Cached("createCountingProfile()") LoopConditionProfile loopProfile) {
+    try {
+      long arraySize = interop.getArraySize(selfArray);
+      loopProfile.profileCounted(arraySize);
+      int[] elemHashCodes = new int[(int) arraySize];
+      for (int i = 0; loopProfile.inject(i < arraySize); i++) {
+        if (interop.isArrayElementReadable(selfArray, i)) {
+          elemHashCodes[i] = (int) hashCodeNode.execute(interop.readArrayElement(selfArray, i));
+        }
+      }
+      return Arrays.hashCode(elemHashCodes);
+    } catch (UnsupportedMessageException | InvalidArrayIndexException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(
+      guards = {"interop.isNull(selfNull)"},
+      limit = "3")
+  long hashCodeForNull(Object selfNull, @CachedLibrary("selfNull") InteropLibrary interop) {
+    return 0;
+  }
+
+  @Specialization(guards = "isHostObject(hostObject)")
+  long hashCodeForHostObject(
+      Object hostObject, @CachedLibrary(limit = "3") InteropLibrary interop) {
+    try {
+      Object hashCodeRes = interop.invokeMember(hostObject, "hashCode");
+      assert interop.fitsInInt(hashCodeRes);
```

Comment:
If it didn't fit `int` it would simply throw `UnsupportedMessageException` which is later propagated as `IllegalStateException`. What's the point of an assert?

Reply:
Right, no point at all. Removing it.